docs/UpgradingToJest30.md

```diff
@@ -0,0 +1,226 @@
+
+### Non-enumerable properties
+
+Non-enumerable object properties are now excluded from object matchers by default. This could affect `expect.objectContaining` or equality checks.
+
+### Improved Type Inference for CalledWith
+
+TypeScript users: The types for the `CalledWith` family of matchers (e.g. `toHaveBeenCalledWith`) have been improved to infer function parameter types. In most cases this will catch type mismatches more accurately. This is a **compile-time** breaking change.
+
+If you were asserting calls with arguments that don’t match the actual function’s parameter types, TypeScript may now error on those tests. For example, if a function is typed to accept a number and you wrote `expect(fn).toHaveBeenCalledWith("string")`, TypeScript 5 + Jest 30’s types will flag this. The runtime behavior of the matcher is unchanged.
+
+To fix new TypeScript errors, ensure your test arguments align with the function’s expected types (or use type casts if you intentionally call with different types).
+
+This change doesn’t impact runtime, but it can surface new type errors in your tests that were previously unnoticed, making your tests more type-safe.
+
+## Jest Mock API Changes
+
+### `jest.genMockFromModule` Removed
+
+The legacy function `jest.genMockFromModule(moduleName)` has been removed. It was previously deprecated in favor of `jest.createMockFromModule(moduleName)`. If you still use `genMockFromModule`, switch to `createMockFromModule` – the behavior is the same. For example:
+
+Old code (Jest 29):
+
+```js
+const mockFs = jest.genMockFromModule('fs');
+```
+
+New code (Jest 30):
+
+```js
+const mockFs = jest.createMockFromModule('fs');
+```
+
+### Removed Mock Function Types
+
+Some TypeScript types related to mock functions have been removed from the public API.
+
+- `MockFunctionMetadata`
+- `MockFunctionMetadataType`
+- `SpyInstance`
+
+If you were using `jest.SpyInstance` in your TypeScript code (for instance, to annotate the return of `jest.spyOn`), you should update to using `jest.Mock` or the more specific `jest.MockedFunction` types.
+
+:::note
+
+These type removals do not affect runtime behavior – they are only relevant for TypeScript users. JavaScript users or tests will not notice any difference.
+
+:::
+
+## Module & Runtime Changes
+
+### ESM Module Support and Internal Restructuring
+
+Jest has introduced significant under-the-hood changes to how its packages are bundled and exported:
+
+- All of Jest’s internal modules are now bundled into single files for faster startup. This means when you install Jest, the number of files it loads is greatly reduced (improving performance). However, a side effect is that any unofficial deep imports into Jest’s packages will likely break. For example, if you previously did something like `require('jest-runner/build/runTest')` (which is not a public API), this path will no longer exist. **Solution:** Use Jest’s public APIs or documented interfaces only.
```

Comment:
Maybe add: "If you are relying on an internal module that you think should be part of the public API, please open a Pull Request"